src/btreemap/node/v2.rs

```diff
@@ -0,0 +1,402 @@
+                next_address.get(),
+            );
+
+            // Write the overflow page contents
+            memory.write(
+                (overflow_addresses[i] + PAGE_OVERFLOW_DATA_OFFSET).get(),
+                &buf[start_idx..end_idx],
+            );
+
+            i += 1;
+        }
+    }
+
+    fn reallocate_overflow_pages<M: Memory>(
+        &self,
+        num_overflow_pages: usize,
+        allocator: &mut Allocator<M>,
+    ) -> Vec<Address> {
+        // Fetch the overflow page addresses of this node.
+        let mut addresses = self.get_overflow_addresses(allocator.memory());
+
+        // If there are too many overflow addresses, deallocate some until we've reached
+        // the number we need.
+        while addresses.len() > num_overflow_pages {
+            allocator.deallocate(addresses.pop().unwrap());
+        }
+
+        // Allocate more pages to accommodate the number requested, if needed.
+        while addresses.len() < num_overflow_pages {
+            addresses.push(allocator.allocate());
+        }
+
+        addresses
+    }
+
+    fn get_overflow_addresses<M: Memory>(&self, memory: &M) -> Vec<Address> {
+        let mut overflow_addresses = vec![];
+        let mut next = self.overflow;
+        while let Some(overflow_address) = next {
+            overflow_addresses.push(overflow_address);
+
+            // Load next overflow address.
+            let maybe_next = Address::from(read_u64(
+                memory,
+                overflow_address + PAGE_OVERFLOW_NEXT_OFFSET,
+            ));
+
+            if maybe_next == crate::types::NULL {
+                next = None;
+            } else {
+                next = Some(maybe_next);
+            }
+        }
+
+        overflow_addresses
+    }
+}
+
+// Reads the entirety of the node, including all its overflows, into a buffer.
+fn read_node<M: Memory>(address: Address, page_size: u32, memory: &M) -> Vec<u8> {
+    // Read the first page of the node.
+    let mut buf = vec![0; page_size as usize];
+    memory.read(address.get(), &mut buf);
+
+    // Append overflow pages, if any.
+    let mut overflow_address = address_from_slice(&buf, OVERFLOW_ADDRESS_OFFSET);
+    while overflow_address != NULL {
+        // Read the overflow.
+        let mut overflow_buf = vec![0; page_size as usize];
+        memory.read(overflow_address.get(), &mut overflow_buf);
+
+        // Validate the magic of the overflow.
+        assert_eq!(&overflow_buf[0..3], OVERFLOW_MAGIC, "Bad magic.");
+
+        // Read the next address
+        overflow_address = address_from_slice(&overflow_buf, PAGE_OVERFLOW_NEXT_OFFSET);
+
+        // Append its data to the buffer.
+        buf.extend_from_slice(&overflow_buf[PAGE_OVERFLOW_DATA_OFFSET.get() as usize..]);
+    }
+
+    buf
+}
+
+fn read_u16_from_slice(slice: &[u8], offset: usize) -> u16 {
+    u16::from_le_bytes((&slice[offset..offset + 2]).try_into().unwrap())
```

Comment:
Perhaps we use `expect` here to be able to provide a somewhat better message?

Reply:
Done.